### ISSUE TITLE
Unvendor grpc/protobuf

### ISSUE THREAD

**Issue by h-vetinari** [conda-forge/ray-packages-feedstock#90]
I didn't realize that ray is [building](https://github.com/ray-project/ray/blob/ray-2.2.0/bazel/ray_deps_build_all.bzl) grpc as a vendored project. This would be a pretty obviously candidate for effing something up when using a newer grpcio.

_Originally posted by @h-vetinari in https://github.com/conda-forge/ray-packages-feedstock/issues/87#issuecomment-1376923344_

**Comment by mattip** [comment]
Do we have any bazel experts around who could either remove the build altogether or figure out how to ignore that error?

**Comment by mattip** [comment]
It seems tensorflow has a [whole scheme](https://github.com/tensorflow/tensorflow/tree/316bbe87c611667c0e0cb0aec525453f953f7835/third_party/systemlibs) to allow using system libraries. Is this [build deps](https://github.com/ray-project/ray/blob/master/bazel/ray_deps_build_all.bzl) the parallel in ray? How would that look for a local grpc?

**Comment by mattip** [comment]
> We use this sort of thing in jaxlib

That passes `TF_SYSTEM_LIBS` down to tensorflow, which has a [whole scheme](https://github.com/tensorflow/tensorflow/tree/316bbe87c611667c0e0cb0aec525453f953f7835/third_party/systemlibs) to allow using system libraries. This mechanism does not exist so far in ray.

**Comment by cread** [comment]
Is anyone still working on this? Having such an old version pinned here is starting to cause some problems for us.

**Comment by mattip** [comment]
Ray 2.4.0 [pins to <1.49](https://github.com/conda-forge/ray-packages-feedstock/blob/8897c18486029ed1daa2a4ac936d2df6e56d4283/recipe/meta.yaml#L82) like [upstream ray](https://github.com/ray-project/ray/blob/cd1ba65e239360c8a7b130f991ed414eccc063ce/python/requirements.txt#L26) on darwin. Would changing to exactly the upstream pinning (`<1.51.3` on non-darwin) help your use-case?

**Comment by cread** [comment]
> Ray 2.4.0 [pins to <1.49](https://github.com/conda-forge/ray-packages-feedstock/blob/8897c18486029ed1daa2a4ac936d2df6e56d4283/recipe/meta.yaml#L82) like [upstream ray](https://github.com/ray-project/ray/blob/cd1ba65e239360c8a7b130f991ed414eccc063ce/python/requirements.txt#L26) on darwin. Would changing to exactly the upstream pinning (`<1.51.3` on non-darwin) help your use-case?

Yes, this would help a lot actually.

**Comment by h-vetinari** [comment]
Good news: dealing with external deps in bazel might _finallyyyyyyy_ be getting easier: https://github.com/conda-forge/tensorflow-feedstock/issues/332

**Comment by mattip** [comment]
It requires bazel 6, which does not seem to work. See ray-project/ray#31504

**Comment by h-vetinari** [comment]
Yeah, but compatibility with modern bazel is mostly just a question of time. The important update here IMO is the new capabilities that'll allow to finally improve the (un)vendoring situation here.

**Comment by anyscalesam** [comment]
@mattip is there anything preventing an update of ray to 2.9.0. that should bring grpcio version to 1.59

**Comment by mattip** [comment]
I found [this article](https://www.stevenengelhardt.com/2021/09/22/practical-bazel-depending-on-a-system-provided-c-cpp-library/) about using native libaries in bazel builds. It seems we could add some patches to replace the grpc and protobuf builds with the conda-provided ones?

**Comment by mattip** [comment]
I think we could try to use these from conda: `libabseil` (instead of `@com_google_absl/*`), 'gprc' (instead of `@com_github_grpc_grpc/*`) 'protobuf' (instead of `@com_google_protobuf`), all from [upstream `BUILD.bazel`](https://github.com/ray-project/ray/blob/master/BUILD.bazel)

**Comment by aslonnie** [comment]
hi, coming from https://github.com/conda-forge/ray-packages-feedstock/issues/253#issuecomment-3695603542

I am trying to understand why unvendoring is a requirement at the first place.

**Comment by mattip** [comment]
Ray uses dynamic linking in _raylet.pyd, so any conda-forge python package that also dynamically links to these libraries should use the same version.
